### PR TITLE
Fix for Issue #77 (does now work with new Google Services API)

### PR DESCRIPTION
**Pull request by Mojo90**
Please check and give Feedback. Does now work for me (tested Adding and Removing Geofences)

### CODE REVIEW

File: plugin.xml [hunk 1]
```diff
@@ -45,7 +43,8 @@
         <source-file src="src/android/Notification.java" target-dir="src/com/cowbell/cordova/geofence" />
         <source-file src="src/android/ReceiveTransitionsIntentService.java" target-dir="src/com/cowbell/cordova/geofence" />
         <source-file src="src/android/RemoveGeofenceCommand.java" target-dir="src/com/cowbell/cordova/geofence" />
-
+		    
+		<framework src="com.google.android.gms:play-services:+" />
```
Comment (tsubik) [inline plugin.xml]: We only need `play-services-location:+` here. Also please ensure using spaces not tabs along the project.

**Comment by tsubik** [comment]
Beside of that small thing I think all is good :+1: Great work. I need to try update google-play-services version on CI to make tests passed. On my devices works well.

**Comment by Mojo90** [comment]
Okay. I handed in my Master Thesis today, so little party tonight and on the weekend I will fix that and make a little cleanup...
Why do you use spaces and not tabs?

**Comment by tsubik** [comment]
Because 4 spaces will always stay as 4 spaces and everything will look consistent everytime. For tabs everyone has different IDE settings, github code diff has different too, etc. :smiley:

**Comment by Hirbod** [comment]
That's true. Spaces FTW

**Comment by Mojo90** [comment]
Okay agree. Another thing what i realised:
If I close my App the geofences are not working. So there is no Service running. As I am not so deep in Android Development I thougt the "ReceiveTransitionsIntentService" is this Service that the App does register Geofence Changes all the time (even when phone restarts). Shouldn't it not be like that or do I have to "fix" this in my App and not related to this plugin?

**Comment by tsubik** [comment]
Hey, everything seems to be fine. I've just checked with my example app. You won't find anything in process/application tab on your phone, this "service" is invoked by system when transition happens. After restarting phone another thing is executed this time BootReceiver - it re-adds all geofences. All services and receivers must be registered in AndroidManifest.xml and the plugin should amend your manifest file https://github.com/cowbell/cordova-plugin-geofence/blob/4ad38ca251c4626d406ce7a49f4709999ef40dac/plugin.xml#L58

File: src/android/GeoNotificationManager.java [hunk 2]
```diff
@@ -12,15 +7,20 @@
 
 import com.google.android.gms.common.ConnectionResult;
 import com.google.android.gms.common.GooglePlayServicesUtil;
+import com.google.android.gms.common.api.GoogleApiClient;
 import com.google.android.gms.location.Geofence;
-import com.google.android.gms.location.LocationClient;
+//import com.google.android.gms.location.LocationClient;
```
Comment (tsubik) [inline src/android/GeoNotificationManager.java]: Ohh, one more thing, as this is no longer used it is better to completely remove commented code.

Comment (Mojo90) [inline src/android/GeoNotificationManager.java]: yep I know that is what I meant with "make a little cleanup"

**Comment by Mojo90** [comment]
Yep I agree with you I thought this is the normal behavior for Geofences. The line you mentioned is correctly added to my Manifest. But if I close the App and change Location (get out of Geofence) nothing happens --> code is not executed / my app does not recognize it. Same behavior if I restart my phone. But I thought it should be as you described. Will check tomorrow and do the cleanup...

**Comment by Mojo90** [comment]
Ok I checked with my S3 Mini.
When I close my app the transition Event is called but Plugin gives an Error that Webview is null.
I think this is another Plugin related bug?
09-06 11:23:12.901  10079-10095/com.develapp.meetfriends D/GeofencePlugin﹕ ReceiveTransitionsIntentService - onHandleIntent
09-06 11:23:12.901  10079-10095/com.develapp.meetfriends D/GeofencePlugin﹕ Geofence transition detected
09-06 11:23:12.961  10079-10095/com.develapp.meetfriends D/GeofencePlugin﹕ Transition Event Received!
09-06 11:23:13.002  10079-10095/com.develapp.meetfriends D/GeofencePlugin﹕ Webview is null

**Comment by tsubik** [comment]
Webview is null is just information, that there is no js context to run code, because your app is closed. All you can do in that situation is set up native notification.

**Comment by Mojo90** [comment]
Ok thanks. will check. I use this https://github.com/katzer/cordova-plugin-local-notifications if geofence transition is detected. Works now in iOS very good ;) but this is another topic...

Do I have to change sth that the changes will not fail?

**Comment by tsubik** [comment]
I think about sending some broadcast message when transition happens, so everyone can receive that message by registered native receiver without messing up with plugin code.

**Comment by tsubik** [comment]
No it's fine, I need to change emulator's image on circle CI to one with newer google play services version installed.

**Comment by Mojo90** [comment]
Okay great, thanks!

One more thing: tried with broadcast message. not able to register javascript method that it listens to a receiver. do you have an example?

**Comment by jgcaruso** [comment]
This is part of an enhancement I meant to do for android (already works in iOS) but I haven't had time lately to work on.




I fully intend to do it at some point though, because I will need the functionality when I catch my android app's functionality up to my iOS one. But if someone else has the knowledge to work on it, feel free.



—

**Comment by jgcaruso** [comment]
But just for some background, if the app isn't open you will not be able to execute a JavaScript method. The function that receives the 'broadcast' will need to be written in device native code.



—

**Comment by Mojo90** [comment]
okay damned. is it maybe possible to "awake/open" app somehow? because I store sth. to javascript localstorage and don't know how to rewrite this in Java.

**Comment by jgcaruso** [comment]
Not that I'm aware of. Imagine how annoying it would be if an app you closed kept opening on you.




Your best bet if you need access to data in the background is to use a plugin to store it natively, and then access it either natively or via JavaScript depending where you are.




Not sure if there are existing plugins for that though. You may need to write your own. 



—

**Comment by Mojo90** [comment]
ok thanks!

**Comment by tsubik** [comment]
Today I've added that broadcasting, how to handle transition in native code described in readme, but as @jgcaruso said there is no easy way to execute javascript if your app is not running (you have to load your app into webview somehow), there might be some workarounds http://stackoverflow.com/questions/18865035/android-using-webview-outside-an-activity-context

**Comment by Mojo90** [comment]
Yes I see, thanks for your support! Will see how to fix that in my App.
I always thought ios is the more "restrictive" one but there Javscript is executed without a problem if app is closed...

**Comment by Mojo90** [comment]
sorry guys I have one more thing. Android does work now very good. 
But my question is now ios related:
2 days ago I said ios is working normal and executing Javascript. So on Simulator if App is closed and Location changes (geofence transition receveid) I get a Notification that geofence transition received. But does not work on real device right now. Is it the same as on Android that Javascript couldn't be executed if App is closed? But why does it work in simulator than? Or is it a bug on my testing device (so far tested only on one iphone 6+) that it should be working on real device as on simulator if App is closed?
Thanks for any help!

**Comment by jgcaruso** [comment]
I'm not an expert on how the plugins work, but I would have assumed that it should NOT work on iOS in the background either, for the exact same reason. The webview does not exist when background events fire.



—

**Comment by andiCR** [comment]
It should work on iPhone, the app loads entirely instead of just an intent.

**Comment by Mojo90** [comment]
okay thanks than it is related to my phone and I need to check with another because as I said on simulator it works just fine :)

**Pull request merged**